src/AWS/CRT/HTTP/Message.php

```diff
@@ -2,32 +2,94 @@
 
 namespace AWS\CRT\HTTP;
 
-abstract class Message {
+use AWS\CRT\NativeResource;
+
+use AWS\CRT\Internal\Encoding;
+
+abstract class Message extends NativeResource {
+    private $method = "GET";
     private $path = "";
     private $query = [];
     private $headers = [];
 
-    public function __construct($path, $query = [], $headers = []) {
+    public function __construct($method, $path, $query = [], $headers = null) {
+        $this->method = $method;
         $this->path = $path;
         $this->query = $query;
-        $this->headers = $headers;
+        $this->headers = !is_null($headers) ? $headers : new Headers();
+        $this->acquire(self::$crt->http_message_new_from_blob(self::marshall($this)));
     }
-}
 
-class Request extends Message {
-    private $body_stream = null;
+    public function __destruct() {
+        self::$crt->http_message_release($this->release());
+        parent::__destruct();
+    }
 
-    public function __construct($path, $query = [], $headers = [], $body_stream = null) {
-        parent::__construct($path, $query, $headers);
-        $this->body_stream = $body_stream;
+    protected static function marshall($msg) {
+        $buf = "";
+        $buf .= Encoding::encodeString($msg->method);
+        $buf .= Encoding::encodeString($msg->pathAndQuery());
+        $buf .= Headers::marshall($msg->headers);
+        return $buf;
     }
-}
 
-class Response extends Message {
-    private $status_code;
+    protected static function _unmarshall($buf, $class=Message::class) {
+        $method = Encoding::readString($buf);
+        $path_and_query = Encoding::readString($buf);
+        if (strchr($path_and_query, "?")) {
+            list($path, $query) = preg_split("/\?/", $path_and_query, 2);
+        } else {
+            list($path, $query) = [$path_and_query, ""];
+        }
+
+        $headers = Headers::unmarshall($buf);
+
+        // Turn query params back into a dictionary
+        if (strlen($query)) {
+            if (strchr($query, "&")) {
```

Comment:
We won't know if the query is currently encoded at this point (or it may have some customer input parameters), so it may be worth calling $query = rawurldecode($query) if it hasn't been done by this point.